packages/kbn-aiops-utils/src/lib/stream_factory.test.ts

```diff
@@ -0,0 +1,181 @@
+  // pattern inside the tests callback to still be able to do async/await for the
+  // `for await()` part. Note that the unzipping here is done just to be able to
+  // decode the stream for the test and assert it. When used in actual code,
+  // the browser on the client side will automatically take care of unzipping
+  // without the need for additional custom code.
+  it('should encode and receive a compressed string based stream', (done) => {
+    (async () => {
+      const { end, push, responseWithHeaders } = streamFactory({
+        'accept-encoding': 'gzip',
+      });
+
+      push('push1');
+      push('push2');
+      end();
+
+      const chunks = [];
+      for await (const chunk of responseWithHeaders.body) {
+        chunks.push(chunk);
+      }
+
+      const buffer = Buffer.concat(chunks);
+
+      zlib.gunzip(buffer, function (err, decoded) {
+        expect(err).toBe(null);
+
+        const streamResult = decoded.toString('utf8');
+
+        expect(responseWithHeaders.headers).toStrictEqual({ 'content-encoding': 'gzip' });
+        expect(streamResult).toBe('push1push2');
+
+        done();
+      });
+    })();
+  });
+
+  it('should encode and receive a compressed NDJSON based stream', (done) => {
+    (async () => {
+      const { DELIMITER, end, push, responseWithHeaders } = streamFactory<MockItem>({
+        'accept-encoding': 'gzip',
+      });
+
+      push(mockItem1);
+      push(mockItem2);
+      end();
+
+      const chunks = [];
+      for await (const chunk of responseWithHeaders.body) {
+        chunks.push(chunk);
+      }
+
+      const buffer = Buffer.concat(chunks);
+
+      zlib.gunzip(buffer, function (err, decoded) {
+        expect(err).toBe(null);
+
+        const streamResult = decoded.toString('utf8');
+
+        const streamItems = streamResult.split(DELIMITER);
+        const lastItem = streamItems.pop();
+
+        const parsedItems = streamItems.map((d) => JSON.parse(d));
+
+        expect(responseWithHeaders.headers).toStrictEqual({ 'content-encoding': 'gzip' });
+        expect(parsedItems).toHaveLength(2);
+        expect(parsedItems[0]).toStrictEqual(mockItem1);
+        expect(parsedItems[1]).toStrictEqual(mockItem2);
+        expect(lastItem).toBe('');
+
+        done();
+      });
+    })();
+  });
+
+  it('should throw when a string based stream receives a non-string chunk', async () => {
+    const { push } = streamFactory({});
+
+    // First push initializes the stream as string based.
+    expect(() => {
+      push('push1');
+    }).not.toThrow();
+
+    // Second push is again a string and should not throw.
+    expect(() => {
+      push('push2');
+    }).not.toThrow();
+
+    // Third push is not a string and should trigger an error.
+    expect(() => {
+      push({ myObject: 'push3' } as unknown as string);
+    }).toThrow('Must not push non-string chunks to a string based stream.');
+  });
+
+  it('should throw when a NDJSON based stream receives a string chunk', async () => {
```

Comment:
Small suggestion to add another simple test case here when the first push is something invalid, like null or  undefined.

Reply:
Added code and test to handle `undefined` with an error here: [4136a92](https://github.com/elastic/kibana/pull/132590/commits/4136a920b096eeb950166515b0ac10b08dd329f6). (I left other values like `null` untouched for now, it might be an edge case but it's a valid JSON value as is)